### ISSUE TITLE
Class-validator vulnerabilities

### ISSUE THREAD

**Issue by msmadja-dev** [typestack/class-validator#1443]
### Description

Recently found for  class-validator critical  vulnerability issue: 
https://snyk.io/vuln/npm:class-validator.

Class-validator is a decorator-based property validation for classes.

Affected versions of this package are vulnerable to Improper Input Validation via bypassing the input validation in validate(), which can lead to cross-site scripting (XSS) or SQL injection. NOTE: There is an optional forbidUnknownValues parameter that can be used to reduce the risk of this bypass. 

**AR**: Found vulnerability when running npm audit. 
**ER**:  No vulnerability found.

**Comment by gokuldeep-kk** [comment]
Facing the same issue on Snyk Test - https://snyk.io/vuln/SNYK-JS-CLASSVALIDATOR-1730566

**Comment by rick-aguayo** [comment]
Any timeframe when this critical vulnerability will be assigned for fix?

**Comment by SandraShklyaeva** [comment]
+1 to have this fix

**Comment by razvanmihaimarin** [comment]
Any updates regarding this?

**Comment by narensgh** [comment]
When are we planning to take this up? This is also throwing vulnerabilities on AWS ECR after enabling image scanning.

**Comment by ChoGathK** [comment]
A cross-site scripting attack occurs when the attacker tricks a legitimate web-based application or site to accept a request as originating from a trusted source.

This is done by escaping the context of the web application; the web application then delivers that data to its users along with other trusted dynamic content, without validating it. The browser unknowingly executes malicious script on the client side (through client-side languages; usually JavaScript or HTML) in order to perform actions that are otherwise typically blocked by the browser’s Same Origin Policy.

Injecting malicious code is the most prevalent manner by which XSS is exploited; for this reason, escaping characters in order to prevent this manipulation is the top method for securing code against this vulnerability.

Escaping means that the application is coded to mark key characters, and particularly key characters included in user input, to prevent those characters from being interpreted in a dangerous context. For example, in HTML, < can be coded as &lt; and > can be coded as &gt; in order to be interpreted and displayed as themselves in text, while within the code itself, they are used for HTML tags. If malicious content is injected into an application that escapes special characters and that malicious content uses < and > as HTML tags, those characters are nonetheless not interpreted as HTML tags by the browser if they’ve been correctly escaped in the application code and in this way the attempted attack is diverted.

The most prominent use of XSS is to steal cookies (source: OWASP HttpOnly) and hijack user sessions, but XSS exploits have been used to expose sensitive information, enable access to privileged services and functionality and deliver malware.

https://security.snyk.io/vuln/SNYK-JS-CLASSVALIDATOR-1730566

**Comment by ChoGathK** [comment]
How to prevent
This section describes the top best practices designed to specifically protect your code:
Sanitize data input in an HTTP request before reflecting it back, ensuring all data is validated, filtered or escaped before echoing anything back to the user, such as the values of query parameters during searches.
Convert special characters such as ?, &, /, <, > and spaces to their respective HTML or URL encoded equivalents.
Give users the option to disable client-side scripts.
Redirect invalid requests.
Detect simultaneous logins, including those from two separate IP addresses, and invalidate those sessions.
Use and enforce a Content Security Policy (source: Wikipedia) to disable any features that might be manipulated for an XSS attack.
Read the documentation for any of the libraries referenced in your code to understand which elements allow for embedded HTML.

**Comment by DnOberon** [comment]
Any updates on this? I really really really don't want to have to rip this library out and figure out a replacement.

**Comment by juansedo** [comment]
Duplicated of #1422.

Any update on this?

**Comment by debragail** [comment]
Hey this is a critical vulnerability can we get some eyes on this?

**Comment by DnOberon** [comment]
honestly @juansedo @debragail I'm thinking this package is dead

**Comment by xujif** [comment]
***@***.***浊酒

**Comment by Shereef** [comment]
For the people asking if there's any update. I don't think there will be.

However,  feel free to use this easy fix:
1. `npm uninstall class-validator`
2. `npm install @nestjs/class-validator`
3. search your project for `from 'class-validator'` and replace it with `from '@nestjs/class-validator'`

P.S. This is what I did and it fits my project and it doesn't necessarily mean that you should do it or that it's right.

**Comment by braaar** [comment]
Duplicate of #1422

**Comment by braaar** [comment]
Closing this. Let's track this in #1422

**Comment by NoNameProvided** [comment]
Hi all!

Sorry for the long overdue update, I would like to chime in to clarify a few things. 

First of all, as mentioned before in other threads the reported issue is not a security vulnerability in the sense that you can defend against it by specifying the `forbidUnknownValues: true` option. 

I still think this was opened by mistake and it's the same as if I would open a vulnerability report on NodeJS saying "if I turn off the server the NodeJS application crashes". The valid (and fixed) problem was in the `class-transformer` package that I fixed in a rather speedy manner in under 10 days from disclosure. (The fastest fix from the 3 affected packages.) That issue was properly reported privately to me with a pentest report explaining the attack vector.

A second problem is that for this vulnerability I was never provided a reproducible test case officially, saying: this is what failing and needs to be fixed. The closest to an official example is from the issue in this repository that is linked in the security reports: #438. Running that example code shows the issue is fixed for almost a year now.

<details>
<summary>Code snippet from #438</summary>

```ts
import { validate, IsInt, Length, IsEmail, IsFQDN, IsDate, Min, Max } from "class-validator"; 
import { plainToClass } from "class-transformer";

class Post {
  @Length(10, 20) 
  title: string; 
  
  @IsInt() 
  @Min(0) 
  @Max(10) 
  rating: number; 
  
  @IsEmail() 
  email: string; 
  
  @IsFQDN() 
  site: string; 
  
  @IsDate() 
  createDate: Date;
}

let userJson = JSON.parse('{"title":1233, "proto":{}}'); // a malformed input 
let users = plainToClass(Post, userJson);

validate(users, { forbidUnknownValues: true }).then(errors => { // errors is an array of validation errors 
  if (errors.length > 0) { 
    console.log("validation failed. errors: ", errors); 
  } else { 
    console.log("validation succeed"); 
  } 
});
```

This code fails with validation errors as expected.

</details>

So you may ask why the security advisory is still open? That's the million-dollar question, and the answer is that when you try to write to someone they will redirect to someone else who will redirect to someone else who will redirect to the first org. It's a circle and everybody says: "sorry I just source my data from someone else", I cannot do anything for you. 

Another contributor tried to write to them, and I have tried to write to them. No success. 

The mistake I made was that after a while I stopped trying. I knew the issue don't exist so I am using it without worrying, but I see how an open critical advisory is scary for others. 

To sum up my plan for going at it again: 
 - I will flip the default settings for `forbidUnknownValues` and make a release so there is literally zero ground to say this is a vulnerability 
 - I will try to contact the SNYK and other organizations again to tell them the issue is not present
 - I will repeat step two until someone actually can update the security advisory 

Also, it is worth noting that the other package under the NestJS org doesn't fix any issues. The security warning is not present there because it has a different name, not because the "problem is fixed".  

PS: As @braaar mentioned, this is tracked in another issue now, please subscribe to that for future changes.
